### PR TITLE
Enhance path validation logic

### PR DESCRIPTION
**Pull request by sebastienros**
Adding query arguments tests for the chars `?`, `/`, `%`, `#`, ` `

Adding these rules for path segment arguments:

- If the string contains '/' or '\' anywhere, fail.
- If after calling string.Trim():
  - The string is empty, fail.
  - The string contains only '.' characters, fail.
  
This is because we don't think it's legitimate for users to be able to use these as valid path arguments since it could be provided by user data (tenant name, username, ...) and escaping could be tricky as servers might un-encode them.

I am not sure `ArgumentException` is the best here so please send suggestions.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/extensions/pull/4495)

### CODE REVIEW

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:47]: Should just use 'string' in the generated code.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:71]: Add blank line before.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:518]: Instead of generating this, could this be a static method in the M.E.EnumStrings package?

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:529]: Is this the best we can do here? This is doing three full iteration of the string. Aren't there some overloads that take an array of chars to find so we only iterate once?

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:554]: Can we avoid this potential allocation?

**Review by geeknoid — COMMENTED** [review]
I'm not sure what we're trying to achieve here. This generator is producing a client library that calls an endpoint. Are we trying to protect the endpoint by preventing the client from sending specific strings? Is this really appropriate? The endpoint is the one that should defend against this, right?

What if the client legitimately needs to send those funny strings?

**Comment by sebastienros** [comment]
> Are we trying to protect the endpoint by preventing the client from sending specific strings?

It's not to protect then endpoint, but to protect the application using the generated code from accessing endpoints that it's not supposed to. The code path it's changing uses method arguments to create the destination endpoint. If this argument is not sanitized then it could be formed in a way that the application would call other endpoints that could potentially be abused, leaking other information from the request, or altering another part of the system.

Simple example: your application intends to send a request to delete a product using `/products/{productId}/delete`, but the user of the app manages to pass a `productId` value of `../users/admin` (this could come from a query string in your own app), which could then delete a user account instead on the destination system.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (danmoseley) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:529]: IndexOfAny.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (danmoseley) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:554]: ROSpan?

File: test/Generators/Microsoft.Gen.AutoClient/Generated/QueryTests.cs
Comment (danmoseley) [inline test/Generators/Microsoft.Gen.AutoClient/Generated/QueryTests.cs:91]: is there any value in a test that puts every other ASCII char in and checks none are escaped?

Comment (GrabYourPitchforks) [reply test/Generators/Microsoft.Gen.AutoClient/Generated/QueryTests.cs:91]: &#x55;&#x6E;&#x6C;&#x69;&#x6B;&#x65;&#x6C;&#x79;&#x2E;&#x20;&#x49;&#x74;&#x27;&#x73;&#x20;&#x70;&#x65;&#x72;&#x66;&#x65;&#x63;&#x74;&#x6C;&#x79;&#x20;&#x76;&#x61;&#x6C;&#x69;&#x64;&#x20;&#x74;&#x6F;&#x20;&#x65;&#x73;&#x63;&#x61;&#x70;&#x65;&#x20;&#x65;&#x76;&#x65;&#x72;&#x79;&#x20;&#x63;&#x68;&#x61;&#x72;&#x61;&#x63;&#x74;&#x65;&#x72;&#x3B;&#x20;&#x74;&#x68;&#x65;&#x20;&#x6F;&#x74;&#x68;&#x65;&#x72;&#x20;&#x65;&#x6E;&#x64;&#x70;&#x6F;&#x69;&#x6E;&#x74;&#x20;&#x77;&#x69;&#x6C;&#x6C;&#x20;&#x64;&#x6F;&#x20;&#x74;&#x68;&#x65;&#x20;&#x72;&#x69;&#x67;&#x68;&#x74;&#x20;&#x74;&#x68;&#x69;&#x6E;&#x67;&#x2E;&#x20;&#x49;&#x6E;&#x20;&#x66;&#x61;&#x63;&#x74;&#x2C;&#x20;&#x65;&#x76;&#x65;&#x72;&#x79;&#x20;&#x73;&#x69;&#x6E;&#x67;&#x6C;&#x65;&#x20;&#x63;&#x68;&#x61;&#x72;&#x61;&#x63;&#x74;&#x65;&#x72;&#x20;&#x6F;&#x66;&#x20;&#x74;&#x68;&#x69;&#x73;&#x20;&#x63;&#x6F;&#x6D;&#x6D;&#x65;&#x6E;&#x74;&#x20;&#x69;&#x73;&#x20;&#x65;&#x73;&#x63;&#x61;&#x70;&#x65;&#x64;&#x20;&#x28;&#x75;&#x73;&#x65;&#x20;&#x74;&#x68;&#x65;&#x20;&#x74;&#x68;&#x72;&#x65;&#x65;&#x20;&#x64;&#x6F;&#x74;&#x73;&#x20;&#x74;&#x6F;&#x20;&#x73;&#x65;&#x65;&#x20;&#x69;&#x74;&#x21;&#x29;&#x2C;&#x20;&#x61;&#x6E;&#x64;&#x20;&#x69;&#x74;&#x20;&#x73;&#x68;&#x6F;&#x75;&#x6C;&#x64;&#x20;&#x72;&#x65;&#x6E;&#x64;&#x65;&#x72;&#x20;&#x63;&#x6F;&#x72;&#x72;&#x65;&#x63;&#x74;&#x6C;&#x79;&#x20;&#x69;&#x6E;&#x20;&#x79;&#x6F;&#x75;&#x72;&#x20;&#x62;&#x72;&#x6F;&#x77;&#x73;&#x65;&#x72;&#x2E;

**Comment by GrabYourPitchforks** [comment]
For context, this is something that came up while threat modeling the AutoClient functionality. When it was presented in the threat modeling session, the scenario provided was that it was trying to simplify the client's ability to invoke these endpoints by allowing the client to perform the equivalent of format string building. Threat modeling asked if it was reasonable to expect that these inputs might come directly from untrusted user input (e.g., provided directly by a remote caller) and thus might be hostile, and the answer was yes.

So we end up with a path traversal-style attack. A malicious client tricks an innocent frontend into making unintended privileged requests to an access-controlled backend. Sebastien talked about this earlier at https://github.com/dotnet/extensions/pull/4495#issuecomment-1743421451. There are two mitigations recommended for this risk.

First, the arguments need to be properly escaped when the URI string is being built. (Sebastien already got to this in some previous PRs.)

Second, certain patterns - even after escaping - may still be dangerous. It's fairly common for web servers to collapse `.` and `..` segments, or to collapse empty segments, or to unescaped `%2F` and treat it as a normal path separator, etc. So if a malicious caller gets one of these patterns into a path segment, it could still lead to path traversal. How you want to approach this depends on the philosophy behind how you're presenting this library to your consumers. If you're positioning this library as something "friendly" that callers don't need much experience to use correctly, then it's prudent to put some defensive checks in place to thwart these types of attacks. If you're positioning this library as something meant for power users and that it's really just a glorified string builder, then you're implicitly expecting your callers to know what values are \[un\]safe to put in any of these segments, and you should document this expectation so that callers _explicitly_ know it.

The security review team was fine with either of the two approaches: defense-in-depth or document warnings for consumers. It's primarily a function of what audience you see consuming this library. Looks like this PR opts for the former.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (danmoseley) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:551]:
```suggestion
var sp = value.AsSpan().Trim();

return (sp.Length != 0 && sp.IndexOf("\\") == -1 && (sp.IndexOf('.') == -1 || sp.IndexOfAnyExcept('.') == -1))
```
would that work? untested

> // If it contains a dot it should only be made of dots if not ending with whitespaces

I read this as, if it contains a dot the dots must be contiguous and be followed by nothing but possibly whitespace?

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: I can do that, the package is quite empty right now, and I thought it was only for enums manipulation (or that it would go away). Does it still make sense and should I then just add a new static class with this helper?

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:551]: > I read this as, if it contains a dot the dots must be contiguous and be followed by nothing but possibly whitespace?

Correct

> would that work? untested

There is no such `ROS<string>.Trim()`, only one that takes a `char/char[]` to be trimmed. But that would make the whole thing much simpler since the complexity is to prevent using `string.Trim()`.

@GrabYourPitchforks would that be ok to check specifically for `new char[] { ' ', '\t', '\n', '\r' };` or a subset of this even?

Comment (GrabYourPitchforks) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:551]: There are other whitespace chars beyond what you listed. For simplicity, if you want to go this route, you should just call the `Trim` API. It's available across all runtimes `ROS<char>` is available on: https://apisof.net/catalog/65462c87-fbb3-e410-a5b0-e57bcdae22f1

Comment (GrabYourPitchforks) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:551]: FWIW, since you're writing a generator, you may need to invoke the extension method by its fully qualified name rather than as an instance method.

```cs
System.ReadOnlySpan<char> trimmed = System.MemoryExtensions.Trim(value.AsSpan());
```

Comment (danmoseley) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:551]: > There is no such ROS<string>.Trim(),

As Mr G. Y. Pitchforks suggested it’s an expansion method (what I pasted does compile)

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: Sorry, I meant a utility method in M.AspNetCore.AutoClient. This has nothing to do with enum strings.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:465]: Why is this public?

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (geeknoid) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:531]: If we assume that 99% of the times, the strings will not contain slashes or dots, can we make a fast trivial exit paths to reduce computation?

if (value.IndexOfAny(SlahesOrDots) < 0)
   return true;

Or something like it.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (dbarbosapn) [inline src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:468]: I'm curious if it's a common pattern across dotnet to make the exception messages as a public constant? Struggling to understand the value...

If it is to allow for testing this exception is thrown, perhaps creating a "AutoClient.RuntimePathValidationException" or something similar would make sense...

**Review by dbarbosapn — CHANGES_REQUESTED** [review]

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:465]: It's part of `private static class {staticsName}` so if it's private the generated code won't be able to access it.

Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs:465]: This could be made internal, though the other properties from this private class are public.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: Moved to `AutoClientUtilities.cs` (new file)

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: `IndexOfAnyExcept` is only .NET7+, so substituted it in other cases.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: These have moved to the utilities class now.

File: src/Generators/Microsoft.Gen.AutoClient/Emitter.cs
Comment (sebastienros) [reply src/Generators/Microsoft.Gen.AutoClient/Emitter.cs]: Done

**Comment by sebastienros** [comment]
Waiting for approvals or feedback

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (joperezr) [inline src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:9]:
```suggestion
namespace Microsoft.Extensions.Http.AutoClient;

/// <summary>
```
NIT: Extra new line causing PR build failure.

Comment (sebastienros) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs]: Feel free to edit the files directly in GH, I won't mind

**Review by joperezr — APPROVED** [review]
Changes look good to me. Thanks for following up on this @sebastienros

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (geeknoid) [inline src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:11]: Let's mark this with EditorBrowsable.Never since it is not intended for use by humans, and only by the code generator.

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (geeknoid) [inline src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:41]:
```suggestion
        // Fast path for most common cases
        var index = trimmed.IndexOfAny(_slashesOrDot);
        if (index == -1)
        {
            return true;
        }

        // Slashes can't be used
        trimmed = trimmed.Slice(index);
        if (trimmed.IndexOfAny(_slashes) >= 0)
        {
            return false;
        }

```

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (sebastienros) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:13]: Done

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (sebastienros) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs]: Had to adapt the suggestion a little so that `trimmed` is not changed or it would have side effects with the last check on dots

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/Microsoft.Extensions.Http.AutoClient.json
Comment (sebastienros) [inline src/Libraries/Microsoft.Extensions.Http.AutoClient/Microsoft.Extensions.Http.AutoClient.json:118]: @geeknoid is that the correct way to change this file (manually)? or should I use a tool for correctness.

Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/Microsoft.Extensions.Http.AutoClient.json:118]: You can change it manually, that's fine.

Normally, I use ./scripts/MakeApiBaselines.ps1 to do this for me.

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (geeknoid) [inline src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:40]: I think you want:

trimmed = trimmed.Slice();

as a separate statement, such that it applies to both this use of IndexOfAny as well as for the following code below

Comment (sebastienros) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:40]: Did that on purpose or a string like `"a.."` would return `false` while it's fine.

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/Microsoft.Extensions.Http.AutoClient.json
Comment (sebastienros) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/Microsoft.Extensions.Http.AutoClient.json:118]: I tried again with ApiChief directly (this ps1 failed with some invalid paths) and it worked. I was previously using it on a code generation dll while it should have been use on the main assembly.
Also added the required `-o` argument to generate the json file.

Example:

```
PS C:\code\extensions\eng\Tools\ApiChief> dotnet run -- "C:\code\extensions\artifacts\bin\Microsoft.Extensions.Http.AutoClient\Debug\net8.0\Microsoft.Extensions.Http.AutoClient.dll" emit baseline -o C:\code\extensions\src\Libraries\Microsoft.Extensions.Http.AutoClient\Microsoft.Extensions.Http.AutoClient.json
```

FYI @joperezr since we were not sure it could output this json file too.

File: src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs
Comment (geeknoid) [reply src/Libraries/Microsoft.Extensions.Http.AutoClient/AutoClientUtilities.cs:40]: Ah, I see.

**Review by geeknoid — APPROVED** [review]

**Review by dbarbosapn — APPROVED** [review]

**Pull request merged**
